Para poder visualizar los mapas se aconseja realizar el siguiente proceso:

1. Abre/ingresa a https://nbviewer.org/
2. En la página principal de nbviewer, verás un campo de texto que dice "Enter the URL of a Notebook in GitHum".
3. Pega la URL en nbviewer
4. Visualiza tu notebook con las gráficas y mapas en nbviewer

In [56]:
 !pip install --upgrade holidays

In [57]:
import pandas as pd
from collections import Counter
import datetime
import holidays

In [58]:
from google.colab import drive

# Mount the Google Drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [59]:
result_df = pd.read_csv('/content/drive/MyDrive/University of Barcelona/Capstone_Project_2024/3 attempt/preprocessed_data_ctxs.csv')
result_df

station_id    year  month   day  hour  percentage_docks_available  \
0               1.0  2020.0    1.0   1.0   3.0                    0.346296   
1               1.0  2020.0    1.0   1.0   8.0                    0.235185   
2               1.0  2020.0    1.0   1.0  13.0                    0.385185   
3               1.0  2020.0    1.0   1.0  18.0                    0.177778   
4               1.0  2020.0    1.0   1.0  23.0                    0.450000   
...             ...     ...    ...   ...   ...                         ...   
2641510       496.0  2023.0   12.0  31.0   3.0                    0.504630   
2641511       496.0  2023.0   12.0  31.0   8.0                    0.388889   
2641512       496.0  2023.0   12.0  31.0  13.0                    0.861111   
2641513       496.0  2023.0   12.0  31.0  18.0                    0.666667   
2641514       496.0  2023.0   12.0  31.0  23.0                    0.717593   

            ctx-4     ctx-3     ctx-2     ctx-1  
0        0.511111  0.492593  0.459259  0.394444  
1        0.283333  0.364815  0.298148  0.233333  
2        0.248148  0.262963  0.342593  0.337037  
3        0.381481  0.335185  0.244444  0.211111  
4        0.079630  0.209259  0.418519  0.492593  
...           ...       ...       ...       ...  
2641510  0.564815  0.518519  0.606481  0.523148  
2641511  0.555556  0.537037  0.490741  0.435185  
2641512  0.490741  0.412037  0.551282  0.646465  
2641513  0.879630  0.819444  0.768519  0.699074  
2641514  0.722222  0.643519  0.736111  0.731481  

[2641515 rows x 10 columns]

Converti a entero las columnas de la fecha

In [60]:
result_df['year'] = result_df['year'].astype(int)
result_df['month'] = result_df['month'].astype(int)
result_df['day'] = result_df['day'].astype(int)

In [61]:
result_df

station_id  year  month  day  hour  percentage_docks_available  \
0               1.0  2020      1    1   3.0                    0.346296   
1               1.0  2020      1    1   8.0                    0.235185   
2               1.0  2020      1    1  13.0                    0.385185   
3               1.0  2020      1    1  18.0                    0.177778   
4               1.0  2020      1    1  23.0                    0.450000   
...             ...   ...    ...  ...   ...                         ...   
2641510       496.0  2023     12   31   3.0                    0.504630   
2641511       496.0  2023     12   31   8.0                    0.388889   
2641512       496.0  2023     12   31  13.0                    0.861111   
2641513       496.0  2023     12   31  18.0                    0.666667   
2641514       496.0  2023     12   31  23.0                    0.717593   

            ctx-4     ctx-3     ctx-2     ctx-1  
0        0.511111  0.492593  0.459259  0.394444  
1        0.283333  0.364815  0.298148  0.233333  
2        0.248148  0.262963  0.342593  0.337037  
3        0.381481  0.335185  0.244444  0.211111  
4        0.079630  0.209259  0.418519  0.492593  
...           ...       ...       ...       ...  
2641510  0.564815  0.518519  0.606481  0.523148  
2641511  0.555556  0.537037  0.490741  0.435185  
2641512  0.490741  0.412037  0.551282  0.646465  
2641513  0.879630  0.819444  0.768519  0.699074  
2641514  0.722222  0.643519  0.736111  0.731481  

[2641515 rows x 10 columns]

Añadi una nueva columna con información sobre "Weekday"

Voy a probar primero con una muestra y luego si funciona lo aplico a toda la base.

In [62]:
year = result_df['year'][0]
month = result_df['month'][0]
day = result_df['day'][0]

In [63]:
date = datetime.date(year, month, day)
date

datetime.date(2020, 1, 1)

In [64]:
weekday = date.weekday()
weekday

2

Weekday va de 0 a 6, siendo 0 el lunes y 6 el domingo.

Segun la libreria.

In [65]:
weekday

2

In [66]:
date.weekday()

2

Voy a aplicar este cálculo a todas las filas de mi Dataframe

In [67]:
# Function to get the weekday
def get_weekday(row):
    date = datetime.date(int(row['year']), int(row['month']), int(row['day']))
    return date.weekday()

# Apply the function to each row and create a new column
result_df['weekday'] = result_df.apply(lambda row: get_weekday(row), axis=1)
result_df

station_id  year  month  day  hour  percentage_docks_available  \
0               1.0  2020      1    1   3.0                    0.346296   
1               1.0  2020      1    1   8.0                    0.235185   
2               1.0  2020      1    1  13.0                    0.385185   
3               1.0  2020      1    1  18.0                    0.177778   
4               1.0  2020      1    1  23.0                    0.450000   
...             ...   ...    ...  ...   ...                         ...   
2641510       496.0  2023     12   31   3.0                    0.504630   
2641511       496.0  2023     12   31   8.0                    0.388889   
2641512       496.0  2023     12   31  13.0                    0.861111   
2641513       496.0  2023     12   31  18.0                    0.666667   
2641514       496.0  2023     12   31  23.0                    0.717593   

            ctx-4     ctx-3     ctx-2     ctx-1  weekday  
0        0.511111  0.492593  0.459259  0.394444        2  
1        0.283333  0.364815  0.298148  0.233333        2  
2        0.248148  0.262963  0.342593  0.337037        2  
3        0.381481  0.335185  0.244444  0.211111        2  
4        0.079630  0.209259  0.418519  0.492593        2  
...           ...       ...       ...       ...      ...  
2641510  0.564815  0.518519  0.606481  0.523148        6  
2641511  0.555556  0.537037  0.490741  0.435185        6  
2641512  0.490741  0.412037  0.551282  0.646465        6  
2641513  0.879630  0.819444  0.768519  0.699074        6  
2641514  0.722222  0.643519  0.736111  0.731481        6  

[2641515 rows x 11 columns]

Agregue una nueva columna con información sobre si es Fin de de Semana

In [68]:
# Function to check if a date is a weekend
def is_weekened(row):
    weekday = row['weekday']
    if weekday == 5 or weekday == 6:
      return True
    return False

# Apply the function to each row and create a new column
result_df['is_weekend'] = result_df.apply(lambda row: is_weekened(row), axis=1)
result_df

station_id  year  month  day  hour  percentage_docks_available  \
0               1.0  2020      1    1   3.0                    0.346296   
1               1.0  2020      1    1   8.0                    0.235185   
2               1.0  2020      1    1  13.0                    0.385185   
3               1.0  2020      1    1  18.0                    0.177778   
4               1.0  2020      1    1  23.0                    0.450000   
...             ...   ...    ...  ...   ...                         ...   
2641510       496.0  2023     12   31   3.0                    0.504630   
2641511       496.0  2023     12   31   8.0                    0.388889   
2641512       496.0  2023     12   31  13.0                    0.861111   
2641513       496.0  2023     12   31  18.0                    0.666667   
2641514       496.0  2023     12   31  23.0                    0.717593   

            ctx-4     ctx-3     ctx-2     ctx-1  weekday  is_weekend  
0        0.511111  0.492593  0.459259  0.394444        2       False  
1        0.283333  0.364815  0.298148  0.233333        2       False  
2        0.248148  0.262963  0.342593  0.337037        2       False  
3        0.381481  0.335185  0.244444  0.211111        2       False  
4        0.079630  0.209259  0.418519  0.492593        2       False  
...           ...       ...       ...       ...      ...         ...  
2641510  0.564815  0.518519  0.606481  0.523148        6        True  
2641511  0.555556  0.537037  0.490741  0.435185        6        True  
2641512  0.490741  0.412037  0.551282  0.646465        6        True  
2641513  0.879630  0.819444  0.768519  0.699074        6        True  
2641514  0.722222  0.643519  0.736111  0.731481        6        True  

[2641515 rows x 12 columns]

Agregue una nueva columna con información sobre si es Festivo

In [69]:
# Function to check if a date is a holiday
def is_holiday(row):
    date = datetime.date(row['year'], row['month'], row['day'])

    # Initialize the holidays library for the given country, spain ES
    holiday_list = holidays.ES()

    return date in holiday_list

# Apply the function to each row and create a new column
result_df['is_festive'] = result_df.apply(lambda row: is_holiday(row), axis=1)
result_df

station_id  year  month  day  hour  percentage_docks_available  \
0               1.0  2020      1    1   3.0                    0.346296   
1               1.0  2020      1    1   8.0                    0.235185   
2               1.0  2020      1    1  13.0                    0.385185   
3               1.0  2020      1    1  18.0                    0.177778   
4               1.0  2020      1    1  23.0                    0.450000   
...             ...   ...    ...  ...   ...                         ...   
2641510       496.0  2023     12   31   3.0                    0.504630   
2641511       496.0  2023     12   31   8.0                    0.388889   
2641512       496.0  2023     12   31  13.0                    0.861111   
2641513       496.0  2023     12   31  18.0                    0.666667   
2641514       496.0  2023     12   31  23.0                    0.717593   

            ctx-4     ctx-3     ctx-2     ctx-1  weekday  is_weekend  \
0        0.511111  0.492593  0.459259  0.394444        2       False   
1        0.283333  0.364815  0.298148  0.233333        2       False   
2        0.248148  0.262963  0.342593  0.337037        2       False   
3        0.381481  0.335185  0.244444  0.211111        2       False   
4        0.079630  0.209259  0.418519  0.492593        2       False   
...           ...       ...       ...       ...      ...         ...   
2641510  0.564815  0.518519  0.606481  0.523148        6        True   
2641511  0.555556  0.537037  0.490741  0.435185        6        True   
2641512  0.490741  0.412037  0.551282  0.646465        6        True   
2641513  0.879630  0.819444  0.768519  0.699074        6        True   
2641514  0.722222  0.643519  0.736111  0.731481        6        True   

         is_festive  
0              True  
1              True  
2              True  
3              True  
4              True  
...             ...  
2641510       False  
2641511       False  
2641512       False  
2641513       False  
2641514       False  

[2641515 rows x 13 columns]

In [70]:
result_df.to_csv('/content/drive/MyDrive/University of Barcelona/Capstone_Project_2024/3 attempt/preprocessed_data_ctxs_weekdays.csv', index=False)

In [71]:
result_df = pd.read_csv('/content/drive/MyDrive/University of Barcelona/Capstone_Project_2024/3 attempt/preprocessed_data_ctxs_weekdays.csv')
result_df

station_id  year  month  day  hour  percentage_docks_available  \
0               1.0  2020      1    1   3.0                    0.346296   
1               1.0  2020      1    1   8.0                    0.235185   
2               1.0  2020      1    1  13.0                    0.385185   
3               1.0  2020      1    1  18.0                    0.177778   
4               1.0  2020      1    1  23.0                    0.450000   
...             ...   ...    ...  ...   ...                         ...   
2641510       496.0  2023     12   31   3.0                    0.504630   
2641511       496.0  2023     12   31   8.0                    0.388889   
2641512       496.0  2023     12   31  13.0                    0.861111   
2641513       496.0  2023     12   31  18.0                    0.666667   
2641514       496.0  2023     12   31  23.0                    0.717593   

            ctx-4     ctx-3     ctx-2     ctx-1  weekday  is_weekend  \
0        0.511111  0.492593  0.459259  0.394444        2       False   
1        0.283333  0.364815  0.298148  0.233333        2       False   
2        0.248148  0.262963  0.342593  0.337037        2       False   
3        0.381481  0.335185  0.244444  0.211111        2       False   
4        0.079630  0.209259  0.418519  0.492593        2       False   
...           ...       ...       ...       ...      ...         ...   
2641510  0.564815  0.518519  0.606481  0.523148        6        True   
2641511  0.555556  0.537037  0.490741  0.435185        6        True   
2641512  0.490741  0.412037  0.551282  0.646465        6        True   
2641513  0.879630  0.819444  0.768519  0.699074        6        True   
2641514  0.722222  0.643519  0.736111  0.731481        6        True   

         is_festive  
0              True  
1              True  
2              True  
3              True  
4              True  
...             ...  
2641510       False  
2641511       False  
2641512       False  
2641513       False  
2641514       False  

[2641515 rows x 13 columns]

Voy a leer el archivo "Informacio_Estacions_Bicing.csv" para obtener los datos de ubicación de las estaciones

In [72]:
filepath_info = '/content/drive/MyDrive/University of Barcelona/Capstone_Project_2024/Informacio_Estacions_Bicing.csv'
df_info = pd.read_csv(filepath_info, sep=',')
df_info

station_id                                 name physical_configuration  \
0             1        GRAN VIA CORTS CATALANES, 760    ELECTRICBIKESTATION   
1             2                C/ ROGER DE FLOR, 126    ELECTRICBIKESTATION   
2             3                        C/ NÀPOLS, 82    ELECTRICBIKESTATION   
3             4                         C/ RIBES, 13    ELECTRICBIKESTATION   
4             5  PG. LLUIS COMPANYS, 11 (ARC TRIOMF)    ELECTRICBIKESTATION   
..          ...                                  ...                    ...   
501         515                    C/ SANT ADRIÀ, 43    ELECTRICBIKESTATION   
502         516                    C/ SANT ADRIÀ, 88    ELECTRICBIKESTATION   
503         517             AV. RASOS DE PEGUERA, 10    ELECTRICBIKESTATION   
504         518                    C/ LLOBREGÓS, 115    ELECTRICBIKESTATION   
505         519                       C/ PEDRELL, 52    ELECTRICBIKESTATION   

           lat       lon  altitude                              address  \
0    41.397978  2.180107      16.0        GRAN VIA CORTS CATALANES, 760   
1    41.395488  2.177198      17.0                C/ ROGER DE FLOR, 126   
2    41.394156  2.181331      11.0                        C/ NÀPOLS, 82   
3    41.393317  2.181248       8.0                         C/ RIBES, 13   
4    41.391103  2.180176       7.0  PG. LLUIS COMPANYS, 11 (ARC TRIOMF)   
..         ...       ...       ...                                  ...   
501  41.435207  2.194800      19.0                    C/ SANT ADRIÀ, 43   
502  41.435460  2.200157      15.0                    C/ SANT ADRIÀ, 88   
503  41.462095  2.178959      44.0             AV. RASOS DE PEGUERA, 10   
504  41.424689  2.157049     112.0                    C/ LLOBREGÓS, 115   
505  41.424655  2.166289     110.0                       C/ PEDRELL, 52   

     post_code  capacity  is_charging_station  nearby_distance  \
0         8013        45                 True           1000.0   
1         8013        29                 True           1000.0   
2         8013        27                 True           1000.0   
3         8013        21                 True           1000.0   
4         8018        39                 True           1000.0   
..         ...       ...                  ...              ...   
501       8030        24                 True           1000.0   
502       8030        21                 True           1000.0   
503       8033        20                 True           1000.0   
504       8032        27                 True           1000.0   
505       8032        24                 True           1000.0   

     _ride_code_support  rental_uris cross_street  
0                  True          NaN          NaN  
1                  True          NaN          NaN  
2                  True          NaN          NaN  
3                  True          NaN          NaN  
4                  True          NaN          NaN  
..                  ...          ...          ...  
501                True          NaN          NaN  
502                True          NaN          NaN  
503                True          NaN          NaN  
504                True          NaN          NaN  
505                True          NaN          NaN  

[506 rows x 14 columns]

Seleccione las columnas de Latitud, Longitud y Codigo Post, que me podrian servir para el analisis que tengo en mente.

In [73]:
df_info[['station_id', 'lat', 'lon', 'post_code']]

station_id        lat       lon  post_code
0             1  41.397978  2.180107       8013
1             2  41.395488  2.177198       8013
2             3  41.394156  2.181331       8013
3             4  41.393317  2.181248       8013
4             5  41.391103  2.180176       8018
..          ...        ...       ...        ...
501         515  41.435207  2.194800       8030
502         516  41.435460  2.200157       8030
503         517  41.462095  2.178959       8033
504         518  41.424689  2.157049       8032
505         519  41.424655  2.166289       8032

[506 rows x 4 columns]

Para poder graficar/dibujar el mapa tuve en cuenta o en consideracion las variables de latitud, longitud y codigo postal de las ubicaciones de las estaciones.

In [74]:
lats = df_info['lat'][:]

In [75]:
lons = df_info['lon'][:]

In [76]:
post_codes = df_info['post_code'][:]

Identique que todas las estaciones estan contenidas en Barcelona.

In [77]:
cities = ['Barcelona'] * len(lats)

Identifique que los valores son constantes pues por ahora no lo voy a usar.

In [78]:
values = [1] * len(lats)

Instale Folium para dibujar mapas

In [79]:
!pip install folium

## 1.  Grafico con los Station_id & Codigo_Postal
(Grafico interactivo)

(Pulsa cualquier punto/Station)

El siguiente codigo es para graficar/dibujar un mapa con la información de la ubicación de las estaciones, adicionando la información sobre el código postal, a la que se puede acceder desde el mapa interactivo.

In [80]:
import folium
import pandas as pd

# Step 1: Create synthetic data for postal codes
data = {
    'postal_code': post_codes,
    'city': cities,
    'value': values,
    'latitude': lats,
    'longitude': lons
}

df = pd.DataFrame(data)

# Step 2: Create a folium map centered around Barcelona
m = folium.Map(location=[41.39, 2.18], zoom_start=12)

# Step 3: Add markers to the map for each postal code
for _, row in df.iterrows():
    folium.CircleMarker(
        location=(row['latitude'], row['longitude']),
        radius=2,
        color='blue',
        fill=True,
        fill_color='blue',
        popup=f"Postal Code: {row['postal_code']}<br>City: {row['city']}<br>Value: {row['value']}"
    ).add_to(m)

# Display the map
m

Luego de verificar que el mapa funciona, voy a tomar los datos de 2023, que es el ano mas actual que tengo. Para realizar un analisis en este ano.

## 2.  Analisis del ano  2023

Solo voy a utilizar la informacion del ano 2023

In [81]:
result_df_2023 = result_df[result_df['year']==2023]
result_df_2023['station_id'] = result_df_2023['station_id'].astype(int)

result_df_2023.head()

<ipython-input-81-67b5865d3946>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  result_df_2023['station_id'] = result_df_2023['station_id'].astype(int)


station_id  year  month  day  hour  percentage_docks_available  \
4995           1  2023      1    1   2.0                    0.161111   
4996           1  2023      1    1   7.0                    0.146296   
4997           1  2023      1    1  12.0                    0.198148   
4998           1  2023      1    1  17.0                    0.364815   
4999           1  2023      1    1  22.0                    0.548148   

         ctx-4     ctx-3     ctx-2     ctx-1  weekday  is_weekend  is_festive  
4995  0.177778  0.185185  0.187037  0.148148        6        True       False  
4996  0.138889  0.146296  0.129630  0.153704        6        True       False  
4997  0.190741  0.222222  0.209259  0.200000        6        True       False  
4998  0.200000  0.300000  0.307407  0.342593        6        True       False  
4999  0.394444  0.283333  0.224074  0.294444        6        True       False

### Filtre los datos con Muelles Disponibles por debajo del 20% para **días entre semana (Lunes a viernes) y NO festivos** .

Guarde las estaciones con ese indicador.

In [82]:
station_ids_weekday_nofestive = result_df_2023[(result_df_2023['is_festive'] == False) & (result_df_2023['is_weekend'] == False) & (result_df_2023['percentage_docks_available'] < 0.2)]['station_id'].values

In [83]:
import numpy as np

station_ids_weekday_nofestive = list(np.unique(station_ids_weekday_nofestive))

In [84]:
len(station_ids_weekday_nofestive)

399

### Filtre los datos con muelles disponibles por debajo del 20% para **días en fin de semana (Sabados y Domingos) o festivos** .

Guarde las estaciones con ese indicador.

In [85]:
station_ids_weekend_festive = result_df_2023[((result_df_2023['is_festive'] == True) | (result_df_2023['is_weekend'] == True)) & (result_df_2023['percentage_docks_available'] < 0.2)]

In [86]:
import numpy as np

station_ids_weekend_festive = list(np.unique(station_ids_weekend_festive))

In [87]:
len(station_ids_weekday_nofestive)

399

Identifico que todas las estaciones (399) están alguna vez en esta situación, seguramente por la recolocación de bicicletas.

## 3.  Analisis de los dias entre Semana (Lunes a viernes) y NO Festivos, sin muelles disponibles, entre 08h00 & 10h00.

In [88]:
station_ids_weekday_nofestive = list(np.unique(result_df_2023[(result_df_2023['is_festive'] == False) & (result_df_2023['is_weekend'] == False) & (result_df_2023['hour'] > 8) & (result_df_2023['hour'] < 10) & (result_df_2023['percentage_docks_available'] == 0.0)]['station_id'].values))
len(station_ids_weekday_nofestive)

45

In [89]:
# Guarde tambien los Station_ids "Comunes" a todos los casos

union_ids = set(station_ids_weekday_nofestive)

In [90]:
df_filtered = df_info[df_info['station_id'].isin(station_ids_weekday_nofestive)]

In [91]:
lats = df_filtered['lat'][:]
lons = df_filtered['lon'][:]
post_codes = df_filtered['post_code'][:]
cities = ['Barcelona'] * len(lats)
values = [1] * len(lats)
ids = df_filtered['station_id'][:]

# Step 1: Create synthetic data for postal codes
data = {
    'postal_code': post_codes,
    'city': cities,
    'value': values,
    'latitude': lats,
    'longitude': lons,
    'ids': ids
}

df = pd.DataFrame(data)

# Step 2: Create a folium map centered around Barcelona
m = folium.Map(location=[41.39, 2.18], zoom_start=12)

# Step 3: Add markers to the map for each postal code
for _, row in df.iterrows():
    folium.CircleMarker(
        location=(row['latitude'], row['longitude']),
        radius=2,
        color='blue',
        fill=True,
        fill_color='blue',
        popup=f"Postal Code: {row['postal_code']}<br>City: {row['city']}<br>Value: {row['value']}<br>Id: {row['ids']}"
    ).add_to(m)

# Display the map
m

## En el mapa podemos localizar las 45 estaciones que se encuentran llenas a primera hora de la mañana en dias laborables 08h00 & 10h00, seguramente coincidiendo con puntos de gran afluencia de empresas.

# 4.  Analisis de los Fines de Semana o Festivos, sin muelles disponibles, entre las 08h00 & 10h00 A.M.

In [92]:
station_ids_weekend_festive = list(np.unique(result_df_2023[((result_df_2023['is_festive'] == True) | (result_df_2023['is_weekend'] == True)) & (result_df_2023['hour'] > 8) & (result_df_2023['hour'] < 10) & (result_df_2023['percentage_docks_available'] == 0)]['station_id'].values))
len(station_ids_weekend_festive)

17

In [93]:
# Guarde tambien los Station_ids "Comunes" a todos los casos

union_ids = set(station_ids_weekend_festive) & union_ids
union_ids

{57, 124, 286}

Estas 3 estaciones tiene problemas de disponibilidad de huecos entre las 08h00 & 10h00, independientemente de si es fin de semana o festivo o no.

In [94]:
df_filtered = df_info[df_info['station_id'].isin(station_ids_weekend_festive)]

In [95]:
lats = df_filtered['lat'][:]
lons = df_filtered['lon'][:]
post_codes = df_filtered['post_code'][:]
cities = ['Barcelona'] * len(lats)
values = [1] * len(lats)
ids = df_filtered['station_id'][:]

# Step 1: Create synthetic data for postal codes
data = {
    'postal_code': post_codes,
    'city': cities,
    'value': values,
    'latitude': lats,
    'longitude': lons,
    'ids': ids
}

df = pd.DataFrame(data)

# Step 2: Create a folium map centered around Barcelona
m = folium.Map(location=[41.39, 2.18], zoom_start=12)

# Step 3: Add markers to the map for each postal code
for _, row in df.iterrows():
    folium.CircleMarker(
        location=(row['latitude'], row['longitude']),
        radius=2,
        color='blue',
        fill=True,
        fill_color='blue',
        popup=f"Postal Code: {row['postal_code']}<br>City: {row['city']}<br>Value: {row['value']}<br>Id: {row['ids']}"
    ).add_to(m)

# Display the map
m


En el mapa podemos localizar las 17 estaciones que en fin de semana o Festivos, tienen las estaciones llenas de bicicletas, sin muelles disponibles, en algun momento entre las 8h00 and 10h00.

Por tanto, considero que no es tan problemático este momento.

Sin embargo, sigo buscando otros escenarios más interesantes para estas fechas.

## 5.  Analisis de los dias Festivos & Fines de Semana, sin muelles disponibles, entre las 12h00 & 18h00 P.M.

In [96]:
station_ids_weekend_festive = list(np.unique(result_df_2023[((result_df_2023['is_festive'] == True) | (result_df_2023['is_weekend'] == True)) & (result_df_2023['hour'] > 12) & (result_df_2023['hour'] < 18) & (result_df_2023['percentage_docks_available'] == 0)]['station_id'].values))
len(station_ids_weekend_festive)

70

In [97]:
# Guarde tambien los Station_ids "Comunes" a todos los casos

union_ids2 = set(station_ids_weekend_festive)

In [98]:
df_filtered = df_info[df_info['station_id'].isin(station_ids_weekend_festive)]

In [99]:
lats = df_filtered['lat'][:]
lons = df_filtered['lon'][:]
post_codes = df_filtered['post_code'][:]
cities = ['Barcelona'] * len(lats)
values = [1] * len(lats)
ids = df_filtered['station_id'][:]

# Step 1: Create synthetic data for postal codes
data = {
    'postal_code': post_codes,
    'city': cities,
    'value': values,
    'latitude': lats,
    'longitude': lons,
    'ids': ids
}

df = pd.DataFrame(data)

# Step 2: Create a folium map centered around Barcelona
m = folium.Map(location=[41.39, 2.18], zoom_start=12)

# Step 3: Add markers to the map for each postal code
for _, row in df.iterrows():
    folium.CircleMarker(
        location=(row['latitude'], row['longitude']),
        radius=2,
        color='blue',
        fill=True,
        fill_color='blue',
        popup=f"Postal Code: {row['postal_code']}<br>City: {row['city']}<br>Value: {row['value']}<br>Id: {row['ids']}"
    ).add_to(m)

# Display the map
m

En el mapa podemos localizar las 70 estaciones que entre los días festivos y fines de semanas entre las 12h00 & 18h00 P.M. Se encuentran llenas. Estas estaciones estan pegadas al paseo marítimo y al borde de la playa, coincidiendo con las horas donde todavia hay claridad (es de dia) para aprovechar de la playa.

## 6.  Analisis de los dias Festivos & Fines de Semana, sin muelles disponibles, entre las 19h00 & 23h00 P.M

In [100]:
station_ids_weekend_festive = list(np.unique(result_df_2023[((result_df_2023['is_festive'] == True) | (result_df_2023['is_weekend'] == True)) & (result_df_2023['hour'] > 19) & (result_df_2023['hour'] < 23) & (result_df_2023['percentage_docks_available'] == 0)]['station_id'].values))
len(station_ids_weekend_festive)

46

In [101]:
# Guarde tambien los Station_ids "Comunes" a todos los casos

union_ids2 = set(station_ids_weekend_festive) & union_ids2
union_ids2

{9,
 55,
 57,
 88,
 104,
 117,
 133,
 147,
 157,
 158,
 171,
 176,
 178,
 179,
 260,
 269,
 279,
 284,
 340,
 348,
 361,
 389,
 392,
 402,
 425}

Y esta lista de estaciones tiene problemas con muelles disponibles entre las 12 y las 6 y entre las 7 y las 23 durante festivos y fines de semana por lo que habría que reforzarlas.

In [102]:
df_filtered = df_info[df_info['station_id'].isin(station_ids_weekend_festive)]

In [103]:
lats = df_filtered['lat'][:]
lons = df_filtered['lon'][:]
post_codes = df_filtered['post_code'][:]
cities = ['Barcelona'] * len(lats)
values = [1] * len(lats)
ids = df_filtered['station_id'][:]

# Step 1: Create synthetic data for postal codes
data = {
    'postal_code': post_codes,
    'city': cities,
    'value': values,
    'latitude': lats,
    'longitude': lons,
    'ids': ids
}

df = pd.DataFrame(data)

# Step 2: Create a folium map centered around Barcelona
m = folium.Map(location=[41.39, 2.18], zoom_start=12)

# Step 3: Add markers to the map for each postal code
for _, row in df.iterrows():
    folium.CircleMarker(
        location=(row['latitude'], row['longitude']),
        radius=2,
        color='blue',
        fill=True,
        fill_color='blue',
        popup=f"Postal Code: {row['postal_code']}<br>City: {row['city']}<br>Value: {row['value']}<br>Id: {row['ids']}"
    ).add_to(m)

# Display the map
m

En el mapa podemos localizar las 46 estaciones, entre las 19h00 y las 23h00, donde estas estaciones del centro parece estar llenas.

## 7.  Analisis sobre el llenado/vaciado de bicicletas, en ciertas estaciones. Los Festivos o Fin de Semana, entre 12h00 & 23h00

Todos estos datos me sirven para verificar/identificar estaciones que se colapsan y habría que ampliar de tamaño o llenar/vaciar de bicicletas con más frecuencia en ciertos tramos horarios.

In [104]:
station_ids_weekend_festive = list(np.unique(result_df_2023[((result_df_2023['is_festive'] == True) | (result_df_2023['is_weekend'] == True)) & (result_df_2023['hour'] > 12) & (result_df_2023['hour'] < 23) & (result_df_2023['percentage_docks_available'] == 0)]['station_id'].values))
len(station_ids_weekend_festive)

101

In [105]:
df_filtered = df_info[df_info['station_id'].isin(station_ids_weekend_festive)]

In [106]:
lats = df_filtered['lat'][:]
lons = df_filtered['lon'][:]
post_codes = df_filtered['post_code'][:]
cities = ['Barcelona'] * len(lats)
values = [1] * len(lats)
ids = df_filtered['station_id'][:]

# Step 1: Create synthetic data for postal codes
data = {
    'postal_code': post_codes,
    'city': cities,
    'value': values,
    'latitude': lats,
    'longitude': lons,
    'ids': ids
}

df = pd.DataFrame(data)

# Step 2: Create a folium map centered around Barcelona
m = folium.Map(location=[41.39, 2.18], zoom_start=12)

# Step 3: Add markers to the map for each postal code
for _, row in df.iterrows():
    folium.CircleMarker(
        location=(row['latitude'], row['longitude']),
        radius=2,
        color='blue',
        fill=True,
        fill_color='blue',
        popup=f"Postal Code: {row['postal_code']}<br>City: {row['city']}<br>Value: {row['value']}<br>Id: {row['ids']}"
    ).add_to(m)

# Display the map
m

En el mapa podemos localizar las 101 estaciones que se encuentran llenas en Festivos y Fines de semana entre las 12h00 y las 23h00,

Mostrando lugares desde donde más gente toma bicicletas para desplazarse en el Fin de Semana.

**La recomendacion seria** que en estas estaciones habria que llenar/surtir con las bicicletas obtenidas anteriormente de **las estaciones llenas** por el Area del Centro o el Area de la Playa.

## 8.  Analisis mapa con información del porcentaje de muelles disponibles de los Festivos o Fin de Semana, entre 12h00 & 23h00

Primero filtre los datos por Festivos o Fin de Semana y en el horario entre las 12h00 & 23h00.

Después saque/identifique/genere la media del valor de la proporción de muelles disponibles en ese tramo horario para cada estación.

In [107]:
result_df_2023_filtered = result_df_2023[((result_df_2023['is_festive'] == True) | (result_df_2023['is_weekend'] == True)) & (result_df_2023['hour'] > 12) & (result_df_2023['hour'] < 23)]
result_df_2023_filtered = result_df_2023_filtered[['station_id', 'percentage_docks_available']].groupby('station_id').mean()
result_df_2023_filtered = result_df_2023_filtered.reset_index()
result_df_2023_filtered

station_id  percentage_docks_available
0             1                    0.769158
1             2                    0.663473
2             3                    0.499445
3             4                    0.496202
4             5                    0.635562
..          ...                         ...
394         474                    0.831421
395         492                    0.536327
396         494                    0.562992
397         495                    0.529004
398         496                    0.574929

[399 rows x 2 columns]

Pege/integre la media de la proporción de muelles disponibles con la información de las estaciones para hacer el mapa.

In [108]:
df_filtered = df_info[df_info['station_id'].isin(list(result_df_2023_filtered['station_id'].values))]

In [109]:
df_merged = pd.merge(df_filtered, result_df_2023_filtered, on='station_id', how='left')

df_merged.head()

station_id                                 name physical_configuration  \
0           1        GRAN VIA CORTS CATALANES, 760    ELECTRICBIKESTATION   
1           2                C/ ROGER DE FLOR, 126    ELECTRICBIKESTATION   
2           3                        C/ NÀPOLS, 82    ELECTRICBIKESTATION   
3           4                         C/ RIBES, 13    ELECTRICBIKESTATION   
4           5  PG. LLUIS COMPANYS, 11 (ARC TRIOMF)    ELECTRICBIKESTATION   

         lat       lon  altitude                              address  \
0  41.397978  2.180107      16.0        GRAN VIA CORTS CATALANES, 760   
1  41.395488  2.177198      17.0                C/ ROGER DE FLOR, 126   
2  41.394156  2.181331      11.0                        C/ NÀPOLS, 82   
3  41.393317  2.181248       8.0                         C/ RIBES, 13   
4  41.391103  2.180176       7.0  PG. LLUIS COMPANYS, 11 (ARC TRIOMF)   

   post_code  capacity  is_charging_station  nearby_distance  \
0       8013        45                 True           1000.0   
1       8013        29                 True           1000.0   
2       8013        27                 True           1000.0   
3       8013        21                 True           1000.0   
4       8018        39                 True           1000.0   

   _ride_code_support  rental_uris cross_street  percentage_docks_available  
0                True          NaN          NaN                    0.769158  
1                True          NaN          NaN                    0.663473  
2                True          NaN          NaN                    0.499445  
3                True          NaN          NaN                    0.496202  
4                True          NaN          NaN                    0.635562

# 9.  Analisis y generacion del mapa con el tamaño de los círculos en función del Inverso del percetage_docks_available.

values = 10*(1-df_merged['percentage_docks_available'][:]) : es la proporción de bicicletas disponibles en la estación.

In [110]:
lats = df_merged['lat'][:]
lons = df_merged['lon'][:]
post_codes = df_merged['post_code'][:]
cities = ['Barcelona'] * len(lats)
values = 10*(1-df_merged['percentage_docks_available'][:])
ids = df_merged['station_id'][:]

# Step 1: Create synthetic data for postal codes
data = {
    'postal_code': post_codes,
    'city': cities,
    'value': values,
    'latitude': lats,
    'longitude': lons,
    'ids': ids
}

df = pd.DataFrame(data)

# Step 2: Create a folium map centered around Barcelona
m = folium.Map(location=[41.39, 2.18], zoom_start=12)

# Step 3: Add markers to the map for each postal code
for _, row in df.iterrows():
    folium.CircleMarker(
        location=(row['latitude'], row['longitude']),
        radius=row['value'],
        color='blue',
        fill=True,
        fill_color='blue',
        popup=f"Postal Code: {row['postal_code']}<br>City: {row['city']}<br>Value: {row['value']}<br>Id: {row['ids']}"
    ).add_to(m)

# Display the map
m

Este último mapa muestra las estaciones según el nivel de saturación durante el día entre 12h00 & 23h00 de días festivos y fines de semanas.

Las estaciones con más saturación tienen círculos más grandes, es decir, aquellas que tienen más bicicletas según capacidad y por tanto menos muelles disponibles.


Claramente se ven las estaciones del Area del Centro y el Area de la Playa, son las más saturadas, en comparacion con el resto de estaciones.

Mariona y Pere, si hacen click en alguna estación del mapa podran obtener/observar/analizar la información sobre esa estación como el Station_id o la proporción de bicicletas disponibles.